### ISSUE TITLE
No way to handle case-transformed XML escapes

### ISSUE THREAD

**Issue by ZevEisenberg** [Raizlabs/BonMot#272]
Let's say you have a string with both XML tags and XML entities, like this. The example is contrived, but it's based on an example I encountered in a real app:

```
"Status: <strong>Expiring in <1 day</strong>"
```

The `<` in the string causes the XML parsing to fail and show the `<strong>` tags in the final string. The solution is to use `&lt;` instead.

However, we were also calling `localizedUppercase` on the string before passing it in to BonMot, transforming it to `"Status: <strong>EXPIRING IN &LT;1 DAY</strong>"` which caused the `&lt;` → `&LT;` to fail to parse. ~I was able to work around it in my app, but it was unexpected and led to weirdness.~ I have not found a clean way to work around it in my app.

We probably want to fix this by just implementing #26 (changing text case) and encourage people to use that. Other approaches, such as making the XML parser case-insensitive, could have performance implications that we can just work around.

**Comment by garnett** [comment]
It seems that I'm facing the same issue with string: "you agree to our <tag>Terms</tag> & <tag>Privacy Policy</tag>"
Bonmot fails to parse "&" and thus show tags in final string.

**Comment by ZevEisenberg** [comment]
@garnett sorry for the inconvenience! Version 4.0 was a breaking change (we use [SemVer](http://semver.org)), but it's still a drag when stuff like this changes. You should be able to work around it by escaping the ampersand as `&amp;`. If that doesn't work for you, please open a new issue with details. I have some hack time today, so I should be able to help.

Thanks for using BonMot, and please let me know how the 3.2 → 4.0 migration goes, and if there's anything I can do to help!

**Comment by garnett** [comment]
@ZevEisenberg escaping character with `&amp;` works, but it isn't really convenient. Almost all of our strings come from `Localizable.strings` and I would prefer to ease lives of our localizators by hiding implementation details.

As a workaround I'm using this function (not really performant and not really accounts for all possible symbols):
```
extension String {
    func xmlEncoded() -> String {
        return self
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&#39;")
    }
}
```

Do you think it makes sense to use similar function inside BonMot?

**Comment by ZevEisenberg** [comment]
@garnett it may indeed make sense. The escaping does add an annoying wrinkle, although if you don't need XML at all in a string, you can avoid it by not including any `.xmlRules` in your `StringStyle`. While I agree about hiding the implementation from localizers, there's going to have to be some sort of middle ground if localizers are going to be able to use XML tags to style substring ranges.

**Comment by garnett** [comment]
My initial use case was not clear enough indeed, sorry.
1. `Localizable.strings` has `You agree to our %@ & %@`
2. At runtime, we use this format to build up the following: `You agree to our <link>Terms</link> & <link>Conditions</link>`.

We have special styling for `<link>` and add tap detectors to the text (not related to the issue).

I think that adding `.escapesXml`  option to `XMLParsingOptions` and have it disabled by default might be good solution here.

**Comment by ZevEisenberg** [comment]
The problem with not escaping XML characters is that a stray `<` or `&` will throw off `XMLParser`. Since we have implemented #26, as this bug's description suggests, I'm going to close this issue. I agree that it's annoying to have to escape characters, but I don't think having an option not to escape them is the right approach. Please feel free to open a new issue if you'd like to discuss this further, because I'm definitely interested in finding ways to make BonMot easier to use and less surprising.

By the way, [this Stack Overflow answer](http://stackoverflow.com/a/2444169/255489) suggests that the only characters that `XMLParser` supports escaping are `&lt;, &gt;, &apos;, &quot; and &amp;`. That suggests that you should be able to use any other UTF-8 characters without needing to escape them.
